[SPARK-28294][CORE] Support `spark.history.fs.cleaner.maxNum` configuration

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

Up to now, Apache Spark maintains the given event log directory by **time** policy, `spark.history.fs.cleaner.maxAge`. However, there are two issues.
1. Some file system has a limitation on the maximum number of files in a single directory. For example, HDFS `dfs.namenode.fs-limits.max-directory-items` is 1024 * 1024 by default.
https://hadoop.apache.org/docs/r3.2.0/hadoop-project-dist/hadoop-hdfs/hdfs-default.xml
2. Spark is sometimes unable to to clean up some old log files due to permission issues (mainly, security policy).

To handle both (1) and (2), this PR aims to support an additional policy configuration for the maximum number of files in the event log directory, `spark.history.fs.cleaner.maxNum`. Spark will try to keep the number of files in the event log directory according to this policy.

## How was this patch tested?

Pass the Jenkins with a newly added test case.

dongjoon-hyun committed:
[SPARK-28294][CORE] Support `spark.history.fs.cleaner.maxNum` configuration

SparkQA commented:
**[Test build #107329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107329/testReport)** for PR 25072 at commit [`59277d6`](https://github.com/apache/spark/commit/59277d698aef2910b8d22ad4b5ef539e959c3769).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12471/
Test PASSed.

SparkQA commented:
**[Test build #107329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107329/testReport)** for PR 25072 at commit [`59277d6`](https://github.com/apache/spark/commit/59277d698aef2910b8d22ad4b5ef539e959c3769).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107329/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -851,10 +836,62 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      oldAttempts.foreach { app =>
+        if (count > 0) {
+          // Applications may have multiple attempts, some of which may not be completed yet.
+          val (toDelete, remaining) = app.attempts.partition { attempt =>
```
`.partition(_.info.completed)`?

dongjoon-hyun replied:
Sure!

srowen commented on core/src/main/scala/org/apache/spark/internal/config/History.scala:
```diff
@@ -49,6 +49,13 @@ private[spark] object History {
 
+  // This is designed to be smaller than the default value of
+  // `dfs.namenode.fs-limits.max-directory-items` (1024 * 1024).
+  val MAX_LOG_NUM = ConfigBuilder("spark.history.fs.cleaner.maxNum")
```
I think this will need to be documented publicly if `spark.history.fs.cleaner.maxAge` is, and document that both limitations apply. Nit: the default isn't quite consistent with the comment above, which says 1024*1024.

dongjoon-hyun replied:
- Sure for the document.
- The value is designed to be smaller as described at line 52 (`This is designed to be smaller than the default value`) because Spark applications will fail due to event log file creation if the directory contains 1024*1024 files.

srowen replied:
OK sounds good.

dongjoon-hyun commented:
Thank you for review, @srowen ! I'll update soon.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -851,10 +836,62 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      val oldAttempts = listing.view(classOf[ApplicationInfoWrapper])
+        .index("oldestAttempt")
+        .asScala
+        .toList
```
The existing one has different order. For example, `expired` is `reversed` and starts from `maxTime`. Do you mean something other?

srowen replied:
I just mean can you remove `.toList`? it's already some kind of Seq

dongjoon-hyun replied:
Oh, I missed this comment.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -851,10 +836,62 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+
+    // If the number of files is bigger than MAX_LOG_NUM,
+    // clean up all completed attempts per application one by one.
+    val num = listing.view(classOf[LogInfo]).index("lastProcessed").asScala.count(_ => true)
```
Yep.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -1066,19 +1103,21 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
-  private def deleteLog(fs: FileSystem, log: Path): Unit = {
+  private def deleteLog(fs: FileSystem, log: Path): Boolean = {
+    var deleted = false
```
We need this because `fs.delete(log, true)` doesn't succeed always.

srowen replied:
OK that's fine

dongjoon-hyun committed:
Address comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12503/
Test PASSed.

SparkQA commented:
**[Test build #107366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107366/testReport)** for PR 25072 at commit [`84ed42b`](https://github.com/apache/spark/commit/84ed42b3f18cff75142563ebca08fffa1e90cc85).

dongjoon-hyun commented:
Could you review once more? I addressed some of the comments. For the followings, please let me know your opinion if I missed the point. I can update more.
- `val oldAttempts = listing.view(classOf[ApplicationInfoWrapper])` (https://github.com/apache/spark/pull/25072#discussion_r301225173)
- `var deleted = false` (https://github.com/apache/spark/pull/25072#discussion_r301226054)

srowen commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -851,10 +836,60 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        }
+      }
+      if (count > 0) {
+        logError(s"Fail to clean up according to MAX_LOG_NUM policy ($maxNum).")
```
One last idea: maybe this should just be a warning. It's not fatal.

dongjoon-hyun replied:
Oh, right. I'll update like that.

srowen commented on docs/monitoring.md:
```diff
@@ -190,7 +190,10 @@ Security options for the Spark History Server are covered more detail in the
     <td>
       How often the filesystem job history cleaner checks for files to delete.
-      Files are only deleted if they are older than <code>spark.history.fs.cleaner.maxAge</code>
+      Files are deleted in two cases. First, they are older than <code>spark.history.fs.cleaner.maxAge</code>.
```
Maybe more like, "Files are deleted if at least one of two conditions holds. First, they're deleted if they're older than ... They are also deleted if the number of files is more than ..."

dongjoon-hyun replied:
Thanks!

srowen reviewed (commented):
Seems reasonable to me. I usually dislike yet another config but I can see the need for this. This default should virtually never affect any cases that work today, right? like if someone had more than a million files it's already probably causing problems? just wanting to make sure this doesn't surprise anyone as a behavior change if possible.

dongjoon-hyun committed:
Address comments

SparkQA commented:
**[Test build #107368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107368/testReport)** for PR 25072 at commit [`3444b73`](https://github.com/apache/spark/commit/3444b7321a44355091ae1dc20397f240b959ef0d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12505/
Test PASSed.

dongjoon-hyun commented:
For the default value, I agree with your concerns.

Although this default value, 1M, is big enough not to surprise general HDFS-default users. But, there might be two exceptions.

1. If they increase `dfs.namenode.fs-limits.max-directory-items` configuration already, the maximum is 6 * 1024 * 1024.
2. If they are using non-HDFS storage. Maybe, S3?

So, not to surprise all users including S3, do you want me to use `Int.MaxValue` instead? I can change like that. Technically, that will disable this feature, but `spark.history.fs.cleaner.enabled` itself is disabled by default, too.

dongjoon-hyun committed:
Use Int.MaxValue by default

dongjoon-hyun commented:
The default value is increased to `Int.MaxValue`, too.

dongjoon-hyun committed:
Remove `.toList`

dongjoon-hyun committed:
fix typo

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12506/
Test PASSed.

SparkQA commented:
**[Test build #107369 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107369/testReport)** for PR 25072 at commit [`f2c93a0`](https://github.com/apache/spark/commit/f2c93a0291877f3e4af03ab279fba93201054d14).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12507/
Test PASSed.

SparkQA commented:
**[Test build #107366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107366/testReport)** for PR 25072 at commit [`84ed42b`](https://github.com/apache/spark/commit/84ed42b3f18cff75142563ebca08fffa1e90cc85).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107366/
Test FAILed.

SparkQA commented:
**[Test build #107369 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107369/testReport)** for PR 25072 at commit [`f2c93a0`](https://github.com/apache/spark/commit/f2c93a0291877f3e4af03ab279fba93201054d14).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107369/
Test FAILed.

SparkQA commented:
**[Test build #107368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107368/testReport)** for PR 25072 at commit [`3444b73`](https://github.com/apache/spark/commit/3444b7321a44355091ae1dc20397f240b959ef0d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107368/
Test PASSed.

dongjoon-hyun commented:
Hi, @srowen .
Could you review this once more please?

srowen reviewed (changes requested):
LGTM. I wonder if @vanzin has any thoughts as he wrote the maxAge cleanup bit.

srowen reviewed (approved):
(Meant to approve)

dongjoon-hyun commented:
Thank you so much for review and approval, @srowen !
Yes. I'll wait for @vanzin 's comment for one day.

Hi, @vanzin . Could you take a look at this if you have some time?

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12553/
Test PASSed.

SparkQA commented:
**[Test build #107417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107417/testReport)** for PR 25072 at commit [`f2c93a0`](https://github.com/apache/spark/commit/f2c93a0291877f3e4af03ab279fba93201054d14).

SparkQA commented:
**[Test build #107417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107417/testReport)** for PR 25072 at commit [`f2c93a0`](https://github.com/apache/spark/commit/f2c93a0291877f3e4af03ab279fba93201054d14).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107417/
Test PASSed.

dongjoon-hyun commented:
Merged to master. Thank you, @srowen .
